File: app/src/main/java/pl/karol202/paintplus/history/Action.java
package pl.karol202.paintplus.history;

import android.graphics.Bitmap;
import pl.karol202.paintplus.image.Image;

public abstract class Action
{
	protected boolean applied;
	private boolean done;
	
	Action()
	{
		applied = false;
		done = true;
	}
	
	boolean undo(Image image)
	{
		if(!done || !applied) return false;
		done = false;
		return true;
	}
	
	boolean redo(Image image)
	{
		if(done || !applied) return false;
		done = true;
		return true;
	}
	
	public void applyAction(Image image)
	{
		image.addHistoryAction(this);
		applied = true;
	}
	
	abstract Bitmap getActionPreview();
	
	abstract int getActionName();
}

File: app/src/main/java/pl/karol202/paintplus/history/action/Action.kt
package pl.karol202.paintplus.history.action

import android.graphics.Bitmap
import androidx.annotation.StringRes
import androidx.core.graphics.applyCanvas
import pl.karol202.paintplus.util.*

sealed class Action
{
	class ToCommit(@StringRes override val name: Int,
	               override val thumbnailBitmap: Bitmap,
	               val commit: () -> ToRevert) : Action()

	class ToRevert(@StringRes override val name: Int,
	               override val thumbnailBitmap: Bitmap,
	               val revert: () -> ToCommit) : Action()

	data class Preset(@StringRes val name: Int,
	                  val previewProvider: () -> Bitmap)
	{
		object Scope
		{
			fun toCommit(action: () -> ToRevert) = action

			fun toRevert(action: () -> ToCommit) = action
		}

		fun commit(action: Scope.() -> () -> ToCommit): ToRevert
		{
			val thumbnail = createThumbnailBitmap(previewProvider())
			return toRevert(thumbnail, Scope.action())
		}

		fun revert(action: Scope.() -> () -> ToRevert): ToCommit
		{
			val thumbnail = createThumbnailBitmap(previewProvider())
			return toCommit(thumbnail, Scope.action())
		}

		fun toCommit(thumbnailBitmap: Bitmap, revert: () -> ToRevert) =
				ToCommit(name, thumbnailBitmap, revert)

		fun toRevert(thumbnailBitmap: Bitmap, commit: () -> ToCommit) =
				ToRevert(name, thumbnailBitmap, commit)
	}

	companion object
	{
		private val maxPreviewSize = squareSize(width = 60)
		private val maxPreviewRect = maxPreviewSize.toRect()

		fun createThumbnailBitmap(bitmap: Bitmap): Bitmap
		{
			val dstRect = bitmap.size.fitInto(maxPreviewSize).toRect().centerInside(maxPreviewRect)
			val dstBitmap = Bitmap.createBitmap(maxPreviewSize.width, maxPreviewSize.height, Bitmap.Config.ARGB_8888)
			return dstBitmap.applyCanvas {
				drawBitmap(bitmap, null, dstRect, null)
			}
		}
	}

	abstract val name: Int
	abstract val thumbnailBitmap: Bitmap
}
